pkg/agent/openflow/pipeline.go

```diff
@@ -768,71 +837,73 @@ func (c *client) l2ForwardCalcFlow(dstMAC net.HardwareAddr, ofPort uint32, skipI
 	// the default flow of L2ForwardingOutTable.
 }
 
-// traceflowL2ForwardOutputFlows generates Traceflow specific flows that outputs traceflow packets to OVS port and Antrea
-// Agent after L2forwarding calculation.
+// traceflowL2ForwardOutputFlows generates Traceflow specific flows that outputs traceflow packets
+// to OVS port and Antrea Agent after L2forwarding calculation.
 func (c *client) traceflowL2ForwardOutputFlows(dataplaneTag uint8, category cookie.Category) []binding.Flow {
 	flows := []binding.Flow{}
-	// Output and SendToController if output port is tunnel or gateway port.
-	// The gw0 IP as Traceflow destination is not supported.
-	if c.encapMode.SupportsEncap() {
-		flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+3).
-			MatchReg(int(PortCacheReg), config.DefaultTunOFPort).
-			MatchIPDscp(dataplaneTag).
-			SetHardTimeout(300).
-			MatchProtocol(binding.ProtocolIP).
-			MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
-			Action().OutputRegRange(int(PortCacheReg), ofPortRegRange).
-			Action().SendToController(uint8(PacketInReasonTF)).
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done())
-		flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+3).
-			MatchReg(int(PortCacheReg), config.DefaultTunOFPort).
+	for _, ipProtocol := range []binding.Protocol{binding.ProtocolIP, binding.ProtocolIPv6} {
+		if c.encapMode.SupportsEncap() {
+			// SendToController and Output if output port is tunnel port.
+			flows = append(flows, c.pipeline[L2ForwardingOutTable].BuildFlow(priorityNormal+3).
+				MatchReg(int(PortCacheReg), config.DefaultTunOFPort).
+				MatchIPDscp(dataplaneTag).
+				SetHardTimeout(300).
+				MatchProtocol(ipProtocol).
+				MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
+				Action().OutputRegRange(int(PortCacheReg), ofPortRegRange).
+				Action().SendToController(uint8(PacketInReasonTF)).
+				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Done())
+			// Only SendToController if output port is local gateway.
```

Comment:
If we still output when kube-proxy is used, could we still trace the Service traffic?

Reply:
In theory, we do not support tracing Service traffic when AntreaProxy is disabled. It is documented (https://github.com/vmware-tanzu/antrea/blob/main/docs/traceflow-guide.md#prerequisites) and the following error will be logged by the agent: `Error: using Service destination requires AntreaProxy feature enabled`. **However, it dates back to before we started using DSCP to carry the Traceflow dataplane tag**. Now that we do, it may be possible to lift the restriction, since iptables will preserve the DSCP field. I commented out the error in the agent code, and this is what I get for a Traceflow request from a Pod to a Service, when AntreaProxy is disabled:

```yaml
name: default-toolbox-jvmv5-to-kube-system-kube-dns-xs25d865
phase: Succeeded
source: default/toolbox-jvmv5
destination: kube-system/kube-dns
results:
- node: k8s-node-worker-1
  timestamp: 1613785297
  observations:
  - component: Forwarding
    componentInfo: Classification
    action: Received
  - component: Forwarding
    componentInfo: Output
    action: Delivered
- node: k8s-node-control-plane
  timestamp: 1613785297
  observations:
  - component: SpoofGuard
    action: Forwarded
  - component: Forwarding
    componentInfo: Output
    action: Forwarded
    tunnelDstIP: 192.168.77.101
- node: k8s-node-control-plane
  timestamp: 1613785297
  observations:
  - component: SpoofGuard
    action: Forwarded
  - component: Forwarding
    componentInfo: Output
    action: Forwarded
```

This is definitely not as straightforward as the one when AntreaProxy is enabled (although it is accurate):
```yaml
name: default-toolbox-jvmv5-to-kube-system-kube-dns-m489kxwk
phase: Succeeded
source: default/toolbox-jvmv5
destination: kube-system/kube-dns
results:
- node: k8s-node-control-plane
  timestamp: 1613785535
  observations:
  - component: SpoofGuard
    action: Forwarded
  - component: LB
    action: Forwarded
    pod: kube-system/coredns-74ff55c5b-vc7z7
    translatedDstIP: 10.10.1.2
  - component: Forwarding
    componentInfo: Output
    action: Forwarded
    tunnelDstIP: 192.168.77.101
- node: k8s-node-worker-1
  timestamp: 1613785535
  observations:
  - component: Forwarding
    componentInfo: Classification
    action: Received
  - component: Forwarding
    componentInfo: Output
    action: Delivered
```

I am personally not sure it is worth lifting the restriction, but I can open a PR to do it and update the documentation. I don't know if there is any edge case I am missing. @jianjuns what do you think?

Reply:
You are right. Probably not worthwhile, given AntreaProxy is enabled by default now.